### PR TITLE
8292285: C2: remove unreachable block after NeverBranch-to-Goto conversion

### PR DESCRIPTION
**Pull request by robcasloz**
`NeverBranch` nodes introduce fake branches out of infinite loops to simplify intermediate optimizations that assume full reachability:
before
`PhaseCFG::convert_NeverBranch_to_Goto()` transforms such nodes into simple `Goto` nodes, removing the fake branch and thus making its destination block unreachable:
after
This changeset removes the unreachable destination block created by `PhaseCFG::convert_NeverBranch_to_Goto()` from the list of basic blocks:
after-removal
Removing unreachable blocks created by `PhaseCFG::convert_NeverBranch_to_Goto()` simplifies late control-flow analysis (for example, recomputing dominators during the output phase as required for late barrier analysis in Generational ZGC) and slightly reduces the size of the generated code.

#### Testing

- hs-tier1-3 (windows-x64, linux-x64, linux-aarch64, and macosx-x64; release and debug mode). The changeset does not introduce new test cases because it is already exercised by at least two existing ones (`compiler/inlining/StringConcatInfiniteLoop.java`, `compiler/loopopts/TestInfiniteLoopNotInnerMost.java`).
- fuzzing (~30 min. on each platform).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292285](https://bugs.openjdk.org/browse/JDK-8292285): C2: remove unreachable block after NeverBranch-to-Goto conversion

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [e633fbec](https://git.openjdk.org/jdk/pull/9853/files/e633fbecde2fc22357394775b70d8ac1ef0196c4)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [e633fbec](https://git.openjdk.org/jdk/pull/9853/files/e633fbecde2fc22357394775b70d8ac1ef0196c4)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9853/head:pull/9853` \
`$ git checkout pull/9853`

Update a local copy of the PR: \
`$ git checkout pull/9853` \
`$ git pull https://git.openjdk.org/jdk pull/9853/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9853`

View PR using the GUI difftool: \
`$ git pr show -t 9853`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9853.diff">https://git.openjdk.org/jdk/pull/9853.diff</a>

</details>

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by robcasloz** [comment]
Thanks for reviewing, Tobias!

**Review by chhagedorn — APPROVED** [review]
Looks good to me, too!

**Comment by robcasloz** [comment]
> Looks good to me, too!

Thanks, Christian!

File: src/hotspot/share/opto/block.cpp [hunk 1]
```diff
@@ -603,6 +603,19 @@ void PhaseCFG::convert_NeverBranch_to_Goto(Block *b) {
   dead->head()->del_req(j);
   for( int k = 1; dead->get_node(k)->is_Phi(); k++ )
     dead->get_node(k)->del_req(j);
+  // If the fake exit block becomes unreachable, remove it from the block list.
```
Comment (vnkozlov) [inline src/hotspot/share/opto/block.cpp:606]: Can you add missing {} in preceding code (lines 593 - 605)?

Comment (robcasloz) [reply src/hotspot/share/opto/block.cpp:611]: Done.

**Comment by robcasloz** [comment]
Thanks for reviewing, Vladimir!

**Pull request closed without merging**